src/appleseed.studio/mainwindow/pythonconsole/pythoneditor.cpp

```diff
@@ -27,50 +27,91 @@
 //
 
 // Interface header.
-#include "pythoninput.h"
+#include "pythoneditor.h"
 
 // appleseed.studio headers.
 #include "pythonhighlighter.h"
+#include "linenumberarea.h"
 
 // Qt headers.
 #include <QKeyEvent>
 #include <QString>
 #include <QStringList>
+#include <QPainter>
 
 // Standard headers.
 #include <string>
 
 namespace appleseed {
 namespace studio {
 
-PythonInput::PythonInput(QWidget* parent)
+PythonEditor::PythonEditor(QWidget* parent)
+  : QPlainTextEdit(parent)
 {
     setUndoRedoEnabled(true);
-    setLineWrapMode(QTextEdit::WidgetWidth);
+    setLineWrapMode(QPlainTextEdit::WidgetWidth);
     setReadOnly(false);
-    setAcceptRichText(false);
     setTextInteractionFlags(
         Qt::TextSelectableByMouse |
         Qt::TextSelectableByKeyboard |
         Qt::TextEditable |
         Qt::TextEditorInteraction);
 
     new PythonSyntaxHighlighter(this->document());
+
+    setStyleSheet("background-color:#303030");
+
+    QFont font("Monospace");
+    font.setStyleHint(QFont::Monospace);
+    setFont(font);
+
+    line_number_area = new LineNumberArea(this);
+
+    connect(this, SIGNAL(cursorPositionChanged()), this, SLOT(slot_highlight_current_line()));
+    slot_highlight_current_line();
+}
+
+void PythonEditor::set_left_margin(int left_margin)
+{
+    setViewportMargins(left_margin, 0, 0, 0);
+}
+
+QTextBlock PythonEditor::get_first_visible_block()
+{
+    return firstVisibleBlock();
+}
+
+int PythonEditor::get_top_of_first_block(QTextBlock block)
+{
+    return (int) blockBoundingGeometry(block).translated(contentOffset()).top();
```

Comment:
Prefer `static_cast<>` whenever possible. C style casts are akin to `reinterpret_cast<>` which are too powerful.